src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Remote/RemoteServiceInvoker.cs

```diff
@@ -91,56 +87,83 @@ public void Dispose()
-            cancellationToken).ConfigureAwait(false);
+    private async static Task<RazorRemoteHostClient> GetMessagePackClientAsync(IWorkspaceProvider workspaceProvider, CancellationToken cancellationToken)
+    {
+        var workspace = workspaceProvider.GetWorkspace();
+
+        var remoteClient = await RazorRemoteHostClient
+            .TryGetClientAsync(
+                workspace.Services,
+                RazorServices.Descriptors,
+                RazorRemoteServiceCallbackDispatcherRegistry.Empty,
+                cancellationToken)
+            .ConfigureAwait(false);
+
+        return remoteClient
+            ?? throw new InvalidOperationException($"Couldn't retrieve {nameof(RazorRemoteHostClient)} for MessagePack serialization.");
+    }
 
-        return remoteClient;
+    private async static Task<RazorRemoteHostClient> GetJsonClientAsync(IWorkspaceProvider workspaceProvider, CancellationToken cancellationToken)
+    {
+        var workspace = workspaceProvider.GetWorkspace();
+
+        var remoteClient = await RazorRemoteHostClient
+            .TryGetClientAsync(
+                workspace.Services,
+                RazorServices.JsonDescriptors,
+                RazorRemoteServiceCallbackDispatcherRegistry.Empty,
+                cancellationToken)
+            .ConfigureAwait(false);
+
+        return remoteClient
+            ?? throw new InvalidOperationException($"Couldn't retrieve {nameof(RazorRemoteHostClient)} for JSON serialization.");
     }
 
-    private async Task<RazorRemoteHostClient?> TryGetJsonClientAsync(CancellationToken cancellationToken)
+    private ValueTask InitializeAsync()
     {
-        var workspace = _workspaceProvider.GetWorkspace();
+        var oopInitialized = _initializeOOPTask is { Status: TaskStatus.RanToCompletion };
+        var lspInitialized = _initializeLspTask is { Status: TaskStatus.RanToCompletion };
 
-        var remoteClient = await RazorRemoteHostClient.TryGetClientAsync(
-            workspace.Services,
-            RazorServices.JsonDescriptors,
-            RazorRemoteServiceCallbackDispatcherRegistry.Empty,
-            cancellationToken).ConfigureAwait(false);
+        // Note: Since InitializeAsync will be called for each remote service call, we provide a synchronous path
+        // to exit quickly when initialized and avoid creating an unnecessary async state machine.
+        return oopInitialized && lspInitialized
+            ? default
+            : new(InitializeCoreAsync(oopInitialized, lspInitialized));
 
-        if (remoteClient is null)
+        async Task InitializeCoreAsync(bool oopInitialized, bool lspInitialized)
         {
-            return null;
-        }
+            // Note: IRemoteClientInitializationService is an IRemoteJsonService, so we always need the JSON client.
+            var remoteClient = await _lazyJsonClient
+                .GetValueAsync(_disposeTokenSource.Token)
+                .ConfigureAwait(false);
```

Comment:
Happy to see `remoteClient` acquisition moving into the method vs. being passed in. Previously it was passed in but only conditionally used based which felt a bit odd. Now it's still conditionally used but the context is much clearer about when that occurs.